src/ripple/app/misc/impl/AMM.cpp

```diff
@@ -0,0 +1,345 @@
+{
+    auto const issues = [&]() -> std::optional<std::pair<Issue, Issue>> {
+        if (optIssue1 && optIssue2)
+            return {{*optIssue1, *optIssue2}};
+        auto const issue1 = ammSle[sfAsset];
+        auto const issue2 = ammSle[sfAsset2];
+        if (optIssue1)
+        {
+            if (*optIssue1 == issue1)
+                return {{issue1, issue2}};
+            else if (*optIssue1 == issue2)
+                return {{issue2, issue1}};
+            JLOG(j.debug()) << "ammHolds: Invalid optIssue1 " << *optIssue1;
+            return std::nullopt;
+        }
+        else if (optIssue2)
+        {
+            if (*optIssue2 == issue2)
+                return {{issue2, issue1}};
+            else if (*optIssue2 == issue1)
+                return {{issue1, issue2}};
+            JLOG(j.debug()) << "ammHolds: Invalid optIssue2 " << *optIssue2;
+            return std::nullopt;
+        }
+        return {{issue1, issue2}};
+    }();
+    if (!issues)
+        return Unexpected(tecAMM_INVALID_TOKENS);
+    auto const [asset1, asset2] = ammPoolHolds(
+        view,
+        ammSle.getAccountID(sfAMMAccount),
+        issues->first,
+        issues->second,
+        j);
+    return std::make_tuple(asset1, asset2, ammSle[sfLPTokenBalance]);
+}
+
+STAmount
+ammLPHolds(
+    ReadView const& view,
+    Currency const& cur1,
+    Currency const& cur2,
+    AccountID const& ammAccount,
+    AccountID const& lpAccount,
+    beast::Journal const j)
+{
+    auto const lptIss = ammLPTIssue(cur1, cur2, ammAccount);
+    return accountHolds(
+        view,
+        lpAccount,
+        lptIss.currency,
+        lptIss.account,
+        FreezeHandling::fhZERO_IF_FROZEN,
+        j);
+}
+
+STAmount
+ammLPHolds(
+    ReadView const& view,
+    SLE const& ammSle,
+    AccountID const& lpAccount,
+    beast::Journal const j)
+{
+    return ammLPHolds(
+        view,
+        ammSle[sfAsset].currency,
+        ammSle[sfAsset2].currency,
+        ammSle[sfAMMAccount],
+        lpAccount,
+        j);
+}
+
+NotTEC
+invalidAMMAsset(
+    Issue const& issue,
+    std::optional<std::pair<Issue, Issue>> const& pair)
+{
+    if (badCurrency() == issue.currency)
+        return temBAD_CURRENCY;
+    if (isXRP(issue) && true != !issue.account)
+        return temBAD_ISSUER;
+    if (pair && issue != pair->first && issue != pair->second)
+        return temBAD_AMM_TOKENS;
+    return tesSUCCESS;
+}
+
+NotTEC
+invalidAMMAssetPair(
+    Issue const& issue1,
+    Issue const& issue2,
+    std::optional<std::pair<Issue, Issue>> const& pair)
+{
+    if (auto const res = invalidAMMAsset(issue1, pair))
+        return res;
+    if (auto const res = invalidAMMAsset(issue2, pair))
+        return res;
+    if (issue1 == issue2)
+        return temBAD_AMM_TOKENS;
+    return tesSUCCESS;
+}
+
+NotTEC
+invalidAMMAmount(
+    std::optional<STAmount> const& a,
+    std::optional<std::pair<Issue, Issue>> const& pair,
+    bool nonNegative)
+{
+    if (!a)
```

Comment:
Not a huge fan of names like "a". Maybe consider a more descriptive name here and elsewhere?